### PR TITLE
[git registry] Add "reference" key

### PR DESCRIPTION
**Pull request by srgbtl**
Fixes: https://github.com/microsoft/vcpkg/issues/19495
Relates to: #99

### CODE REVIEW

File: src/vcpkg/registries.cpp [hunk 1]
```diff
@@ -1135,14 +1146,21 @@ namespace vcpkg
...
-    LockFile::Entry LockFile::get_or_fetch(const VcpkgPaths& paths, StringView key)
+    LockFile::Entry LockFile::get_or_fetch(const VcpkgPaths& paths, StringView repo, StringView reference)
     {
-        auto it = lockdata.find(key);
+        auto lockdata_key = repo.to_string();
+        if (reference != "HEAD")
+        {
+            // old-style, only repo. implicitly HEAD
+            Strings::append(lockdata_key, "@", reference);
+        }
```
Comment (strega-nil-ms) [inline src/vcpkg/registries.cpp:1156]:
```suggestion
        auto lockdata_key = Strings::concat(repo, "@", reference);
```

We always want to use the `@` syntax from now on, and update old fields when we see them.

File: src/vcpkg/registries.cpp [hunk 2]
```diff
@@ -1135,14 +1146,21 @@ namespace vcpkg
...
+
+        auto it = lockdata.find(lockdata_key);
```
Comment (strega-nil-ms) [inline src/vcpkg/registries.cpp:1158]:
```suggestion
        auto it = lockdata.find(lockdata_key);
        if (it == lockdata.end() && reference == "HEAD")
        {
            it = lockdata.find(StringView(lockdata_key).substr(0, lockdata_key.size() - 5));
        }
```

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]

**Review by strega-nil-ms — APPROVED** [review]
Still needs review from @ras0219-msft

File: src/vcpkg/registries.cpp (hunk 2 above)
Comment (ras0219-msft) [reply src/vcpkg/registries.cpp:1153]: I don't think this is necessary -- in fact to minimize the possibility of collisions with usernames, I think we should not have this fallback.

File: src/vcpkg/registries.cpp [hunk 3]
```diff
@@ -1151,9 +1168,18 @@ namespace vcpkg
     {
         if (data->second.stale)
         {
-            print2("Fetching registry information from ", data->first, "...\n");
-            data->second.value =
-                paths.git_fetch_from_remote_registry(data->first, "HEAD").value_or_exit(VCPKG_LINE_INFO);
+            StringView lockdata_key(data->first);
+            print2("Fetching registry information from ", lockdata_key, "...\n");
+
+            auto at_position = lockdata_key.end() - std::find(lockdata_key.begin(), lockdata_key.end(), '@');
```
Comment (ras0219-msft) [inline src/vcpkg/registries.cpp:1174]: This does not work if the url has an `@` in it, such as `https://git@github.com`. There's no reason to inflict this kind of parsing pain on ourselves; we should store these strings separately.

File: azure-pipelines/end-to-end-tests-dir/registries.ps1 [hunk 4]
```diff
@@ -201,7 +201,7 @@ try
 
     Run-Vcpkg install @builtinRegistryArgs '--feature-flags=registries,manifests'
     Throw-IfFailed
-    Require-FileEquals $installRoot/vcpkg/vcpkg-lock.json "{`n  $(ConvertTo-Json $gitRegistryUpstream): `"$gitBaselineCommit`"`n}`n"
+    Require-FileEquals $installRoot/vcpkg/vcpkg-lock.json "{`n  $(ConvertTo-Json "$gitRegistryUpstream@HEAD"): `"$gitBaselineCommit`"`n}`n"
```
Comment (ras0219-msft) [inline azure-pipelines/end-to-end-tests-dir/registries.ps1:204]: This is what I would suggest changing the file format to:

```suggestion
    Require-FileEquals $installRoot/vcpkg/vcpkg-lock.json "{`n  $(ConvertTo-Json "$gitRegistryUpstream"): {`n    `"HEAD`": `"$gitBaselineCommit`"`n  }`n}`n"
```

File: azure-pipelines/end-to-end-tests-dir/registries.ps1 [hunk 5]
```diff
@@ -201,7 +201,7 @@ try
 
     Run-Vcpkg install @builtinRegistryArgs '--feature-flags=registries,manifests'
```
Comment (ras0219-msft) [inline azure-pipelines/end-to-end-tests-dir/registries.ps1:202]: We also need to add a test that actually ensures the reference is put into the lock json.

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
`@`s are a supported character in git remote urls

File: azure-pipelines/end-to-end-tests-dir/registries.ps1 (hunk 4 above)
Comment (srgbtl) [reply azure-pipelines/end-to-end-tests-dir/registries.ps1]: Done.

File: src/vcpkg/registries.cpp [hunk 6]
```diff
@@ -1130,25 +1142,34 @@ namespace vcpkg
...
-    LockFile::Entry LockFile::get_or_fetch(const VcpkgPaths& paths, StringView key)
+    LockFile::Entry LockFile::get_or_fetch(const VcpkgPaths& paths, StringView repo, StringView reference)
     {
-        auto it = lockdata.find(key);
+        auto range = lockdata.equal_range(repo);
+        auto it = std::find_if(range.first, range.second, [&reference](const LockDataType::value_type& repo2entry) {
+            return repo2entry.second.reference == reference;
+        });
+
         if (it == lockdata.end())
```
Comment (ras0219-msft) [inline src/vcpkg/registries.cpp:1152]: Should this be:
```suggestion
        if (it == range.second)
```

**Review by ras0219-msft — COMMENTED** [review]
This LGTM other than my one comment. Thanks for the great work!

**Review by ras0219-msft — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for the new key!

**Pull request merged**